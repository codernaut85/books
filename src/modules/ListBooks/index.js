import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { 
  InputGroup,
  FormControl,
  Spinner, 
  Table 
} from "react-bootstrap";
import { useParams } from "react-router-dom";
import bookActions from '../../store/actions/books/';
import PaginationComponent from '../../components/Pagination';

function renderEmptyBookList() {
  return <div className="book-list__no-books"><p>No books found</p></div>
}

function renderLoading() {
  return <div className="book-list__loading">
    <Spinner animation="border" role="status">
      <span className="sr-only">Loading...</span>
    </Spinner>
  </div>
}

function renderBookList(bookList = []) {
  return (<Table striped bordered hover className="book-list__table">
    <thead>
      <tr>
        <th>ID</th>
        <th>Author</th>
        <th>Title</th>
        <th>Publication Year</th>
        <th>Publication Country</th>
        <th>Publication City</th>
        <th>Pages</th>
      </tr>
    </thead>
    <tbody>
      {
        bookList.map((book, idx) => (<tr key={idx} className="books-table__row">
          <td>{book.id}</td>
          <td>{book.book_author}</td>
          <td>{book.book_title}</td>
          <td>{book.book_publication_year}</td>
          <td>{book.book_publication_country}</td>
          <td>{book.book_publication_city}</td>
          <td>{book.book_pages}</td>
        </tr>))
      }
    </tbody>
  </Table>);
}

function ListBooks() {
  const { page } = useParams();

  const dispatch = useDispatch();

  const { books, count, isLoading } = useSelector(state => state.books);

  const [searchFilter, setSearchFilter] = useState("");

  useEffect(() => {
    dispatch(bookActions.getBooks({
      page: Number(page),
      filters: [{ type: "all", values: [searchFilter] }]
    }));
  }, [dispatch, page, searchFilter]);

  const hasBooks = count > 0;

  return <section className="wrapper book-list">
    <h2 className="heading-level-two">Book Archives</h2>

    <div className="row">
      <InputGroup className="mb-3 col-md-6">
        <InputGroup.Prepend>
          <InputGroup.Text id="basic-addon1">Filter books by wildcard</InputGroup.Text>
        </InputGroup.Prepend>
        <FormControl
          placeholder=""
          aria-label="filter"
          aria-describedby="basic-addon1"
          onChange={e => setSearchFilter(e.target.value)}
          value={searchFilter}
        />
      </InputGroup>
    </div>

    {hasBooks && renderBookList(books)}
    {isLoading && renderLoading()}
    {(!isLoading && !hasBooks) && renderEmptyBookList()}
    {hasBooks && <PaginationComponent page={Number(page)} bookCount={count} />}
  </section>
};

export default ListBooks;