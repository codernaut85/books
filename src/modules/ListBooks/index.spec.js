import TestRenderer from 'react-test-renderer';
import ListBooks from './index';
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';
import { mockBookData } from '../../fixtures';
import PaginationComponent from '../../components/Pagination';

const mockStore = configureStore([]);

describe("ListBooks", () => {  
  it("renders correctly when there are no books in the store", () => {
    const store = mockStore({
      books: {
        books: [],
        count: 0
      }
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);
    
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it("renders correctly when there are some books in the store", () => {
    const store = mockStore({
      books: mockBookData
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);
    
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it("renders the 'no books' content when there are no books", () => {
    const store = mockStore({
      books: {
        books: [],
        count: 0
      }
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);

    const testInstance = testRenderer.root;
    const noBooksContent = testInstance.findByProps({ className: "book-list__no-books" });

    expect(noBooksContent).toBeTruthy();
  });

  it("renders as many rows in the books table as there are books in the store", () => {
    const store = mockStore({
      books: mockBookData
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);

    const testInstance = testRenderer.root;
    const tableRows = testInstance.findAllByProps({ className: "books-table__row" });

    const amountOfBooks = store.getState().books.books.length;
    expect(tableRows.length).toEqual(amountOfBooks);
  });

  it("renders the pagination controls when there are books to show", () => {
    const store = mockStore({
      books: mockBookData
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);

    const testInstance = testRenderer.root;

    expect(testInstance.findByType(PaginationComponent)).toBeTruthy();
    expect(testInstance.findByType(PaginationComponent).props.bookCount).toBe(20);
  });

  it("does not render the pagination controls when there are no books to show", () => {
    const store = mockStore({
      books: {
        books: [],
        count: 0
      }
    });

    const testRenderer = TestRenderer.create(<Router>
      <Provider store={store}>
        <ListBooks />
      </Provider>
    </Router>);

    const testInstance = testRenderer.root;

    const func = () => testInstance.findByType(PaginationComponent);
    expect(func).toThrowError("No instances found with node type: \"PaginationComponent\"");
  });

});