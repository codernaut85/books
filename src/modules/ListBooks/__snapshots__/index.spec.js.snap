// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBooks renders correctly when there are no books in the store 1`] = `
<section
  className="wrapper book-list"
>
  <h2
    className="heading-level-two"
  >
    Book Archives
  </h2>
  <div
    className="row"
  >
    <div
      className="mb-3 col-md-6 input-group"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="basic-addon1"
        >
          Filter books by wildcard
        </span>
      </div>
      <input
        aria-describedby="basic-addon1"
        aria-label="filter"
        className="form-control"
        onChange={[Function]}
        placeholder=""
        value=""
      />
    </div>
  </div>
  <div
    className="book-list__no-books"
  >
    <p>
      No books found
    </p>
  </div>
</section>
`;

exports[`ListBooks renders correctly when there are some books in the store 1`] = `
<section
  className="wrapper book-list"
>
  <h2
    className="heading-level-two"
  >
    Book Archives
  </h2>
  <div
    className="row"
  >
    <div
      className="mb-3 col-md-6 input-group"
    >
      <div
        className="input-group-prepend"
      >
        <span
          className="input-group-text"
          id="basic-addon1"
        >
          Filter books by wildcard
        </span>
      </div>
      <input
        aria-describedby="basic-addon1"
        aria-label="filter"
        className="form-control"
        onChange={[Function]}
        placeholder=""
        value=""
      />
    </div>
  </div>
  <table
    className="book-list__table table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          ID
        </th>
        <th>
          Author
        </th>
        <th>
          Title
        </th>
        <th>
          Publication Year
        </th>
        <th>
          Publication Country
        </th>
        <th>
          Publication City
        </th>
        <th>
          Pages
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="books-table__row"
      >
        <td>
          2086
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1529
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          104
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2060
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των ρωμαίων μεγάλου Βελισαρίου
        </td>
        <td>
          1548
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          32
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2087
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1553
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          104
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2061
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των Ρωμαίων μεγάλου Βελισαρίου
        </td>
        <td>
          1554
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          0
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2062
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των ρωμαίων μεγάλου Βελισαρίου
        </td>
        <td>
          1562
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          32
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2063
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των ρωμαίων μεγάλου Βελισαρίου
        </td>
        <td>
          1567
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          28
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2065
        </td>
        <td>
          Αχέλης, Αντώνιος
        </td>
        <td>
          Βιβλίον --- περιέχον της Μάλτας πολιορκίαν
        </td>
        <td>
          1571
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          84
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2064
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Διήγησις εις τας πράξεις του περιβοήτου στρατηγού των ρωμαίων μεγάλου Βελισαρίου
        </td>
        <td>
          1577
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          32
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2088
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1600
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          92
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2089
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1603
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          92
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2066
        </td>
        <td>
          Πολίτης, Ματθαίος
        </td>
        <td>
          Βίος Δημητρίου Βασιλέως Μοσχοβίας
        </td>
        <td>
          1612
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          24
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2079
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          Πόνησις Χριστοφόρου του Αγγέλου Έλληνος του πολλών πληγών και Μαστίγων γευσαμένου αδίκως παρά των Τούρκων διά την εις Χριστόν Πίστιν
        </td>
        <td>
          1617
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Οξφόρδη
        </td>
        <td>
          12
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2080
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          A Grecian who tasted of many stripes and torments inflicted by the Turkes for the faith
        </td>
        <td>
          1617
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Οξφόρδη
        </td>
        <td>
          16
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2076
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          A Grecian who tasted of many stripes and torments inflicted by the Turkes for the faith
        </td>
        <td>
          1618
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Οξφόρδη
        </td>
        <td>
          16
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2069
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          Εγχειρίδιον περί της καταστάσεως των σήμερον ευρισκομένων Ελλήνων
        </td>
        <td>
          1619
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Καίμπριτζ
        </td>
        <td>
          70
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2075
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          Enchiridion de institutis Graecorum
        </td>
        <td>
          1619
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Καίμπριτζ
        </td>
        <td>
          64
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2084
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          A Grecian who tasted of many stripes and torments inflicted by the Turkes for the faith
        </td>
        <td>
          1620
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Λονδίνο
        </td>
        <td>
          18
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2090
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1620
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          126
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2083
        </td>
        <td>
          Άγγελος, Χριστόφορος (†1638)
        </td>
        <td>
          Of the conditions of the life in which the Greekes now live
        </td>
        <td>
          1625
        </td>
        <td>
          Αγγλία
        </td>
        <td>
          Λονδίνο
        </td>
        <td>
          10
        </td>
      </tr>
      <tr
        className="books-table__row"
      >
        <td>
          2091
        </td>
        <td>
          Ανώνυμος
        </td>
        <td>
          Ο Αλέξανδρος ο Μακεδών
        </td>
        <td>
          1630
        </td>
        <td>
          Ιταλία
        </td>
        <td>
          Βενετία
        </td>
        <td>
          126
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="pagination-controls"
  >
    <ul
      className="pagination"
    >
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/books/page/1"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <span
            aria-hidden="true"
          >
            «
          </span>
          <span
            className="sr-only"
          >
            First
          </span>
        </a>
      </li>
      <li
        className="page-item"
      >
        <a
          className="page-link"
          href="/books/page/1"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <span
            aria-hidden="true"
          >
            »
          </span>
          <span
            className="sr-only"
          >
            Last
          </span>
        </a>
      </li>
    </ul>
  </div>
</section>
`;
